doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -1,209 +1,213 @@
+   .. coqdoc::
 
-.. tacn:: rewrite @term
-   :name: rewrite
+      Inductive eq {A : Type} (x : A) : A -> Prop := eq_refl : eq x x.
 
-   This tactic applies to any goal. The type of :token:`term` must have the form
+   The notation `x = y` is equivalent to `eq x y`.  The notation `x = y :> A`
+   gives the type of x and y explicitly.
 
-   ``forall (x``:sub:`1` ``:A``:sub:`1` ``) ... (x``:sub:`n` ``:A``:sub:`n` ``), eq term``:sub:`1` ``term``:sub:`2` ``.``
+- :gdef:`Setoid equality <setoid equality>` defines equality in terms of an equivalence
+  relation.  A :gdef:`setoid` is a set that is equipped with an equivalence relation
+  (see https://en.wikipedia.org/wiki/Setoid).  These are needed to form a :gdef:`quotient set`
+  or :gdef:`quotient`
+  (see https://en.wikipedia.org/wiki/Equivalence_Class).  In Coq, users generally work
+  with setoids rather than constructing quotients, for which there is no specific support.
 
-   where :g:`eq` is the Leibniz equality or a registered setoid equality.
+- :gdef:`Definitional equality <definitional equality>` is equality based on the conversion
+  rules, which Coq can determine automatically.  (See :ref:`Conversion-rules`)
+  When two terms are definitionally equal, Coq knows it can
+  replace one with the other, such as with :tacn:`change` `X with Y`, among many
+  other advantages.  ":term:`Convertible <convertible>`" is another way of saying that
+  two terms are definitionally equal.
 
-   Then :n:`rewrite @term` finds the first subterm matching `term`\ :sub:`1` in the goal,
-   resulting in instances `term`:sub:`1`' and `term`:sub:`2`' and then
-   replaces every occurrence of `term`:subscript:`1`' by `term`:subscript:`2`'.
-   Hence, some of the variables :g:`x`\ :sub:`i` are solved by unification,
-   and some of the types :g:`A`\ :sub:`1`:g:`, ..., A`\ :sub:`n` become new
-   subgoals.
+  .. todo PR: A and B are convertible iff they are definitionally equal??
 
-   .. exn:: The @term provided does not end with an equation.
-      :undocumented:
+.. _rewritingexpressions:
 
-   .. exn:: Tactic generated a subgoal identical to the original goal. This happens if @term does not occur in the goal.
-      :undocumented:
+Rewriting with Leibniz and setoid equality
+------------------------------------------
 
-   .. tacv:: rewrite -> @term
+.. tacn:: rewrite {+, @oriented_rewriter } {? @occurrences } {? by @ltac_expr3 }
 
-      Is equivalent to :n:`rewrite @term`
+   .. insertprodn oriented_rewriter one_term_with_bindings
 
-   .. tacv:: rewrite <- @term
+   .. prodn::
+      oriented_rewriter ::= {? {| -> | <- } } {? @natural } {? {| ? | ! } } @one_term_with_bindings
+      one_term_with_bindings ::= {? > } @one_term {? with @bindings }
 
-      Uses the equality :n:`@term`:sub:`1` :n:`= @term` :sub:`2` from right to left
+   .. todo PR: When I search for "rewrite (" in the .v files, I see tons of examples that I don't
+      understand.  Need to say something about them.
```

Comment:
Yes, the current text is not explaining that properly. You can give the name of a lemma or hypothesis (this is the most common case) but you can also give any term whose type is of the right form (shown below). The typical use case will be to apply a lemma to its argument explicitly (also the `with` clause also supports this).

Reply:
Getting there.  Explain me these:

```
rewrite (not_not_iff P) in H by dec
rewrite (add_lt_mono_r _ _ n)          (is there an implicit parm?)
rewrite (CR_of_Q_plus R 1 1)           (also implicit?)
```

Maybe we should mention `_` as a wildcard, too?

Reply:
If you provide a term, do you have the prove the equality in a side condition (side goal)?

Reply:
> Getting there. Explain me these:
> 
> ```
> rewrite (not_not_iff P) in H by dec
> rewrite (add_lt_mono_r _ _ n)          (is there an implicit parm?)
> rewrite (CR_of_Q_plus R 1 1)           (also implicit?)
> ```

It's not clear what you are asking, but basically, we pass some parameters to these lemmas (which Coq may or may not have inferred by itself). The `_` wildcard allows skipping some parameters which we know Coq will correctly infer.

> Maybe we should mention `_` as a wildcard, too?

It's not specific to `rewrite` though.

> If you provide a term, do you have the prove the equality in a side condition (side goal)?

No. It works exactly the same as if you provide a hypothesis or a lemma (which are terms too). The type of the term must have the expected form (i.e. `forall ..., EQ term1 term2`). The proof of the equality is the term itself (a proof term).

Reply:
The examples are not in the form we described, `(EQ term1 term2)`.  The first has only one parameter, the others have three.  What do these examples mean and how are they interpreted?  Is there some kind of evaluation applied before doing the rewrite?

Reply:
> The examples are not in the form we described, `(EQ term1 term2)`. The first has only one parameter, the others have three.

It's not the provided term but its type which must have the form `forall ..., EQ term1 term2`. This is the case in these examples. This is exactly the same as when giving lemma or hypothesis names (it's the type of `H` which must be in the form `forall ..., EQ term1 term2`).

Reply:
That needs some explanation.  It's not at all obvious and it's new to me.  Dropping in "type of a term" is just hiding the details.  How many tactics rely on the type of the term?

Reply:
> How many tactics rely on the type of the term?

The answer is basically all. This is very fundamental stuff so very important that you understand fully. We can move the discussion somewhere else (or have a call) if needed.

I'll try to explain once more using `rewrite` and `apply` as examples:

If we forget the (non-important here) considerations on opaque vs transparent definitions, a lemma and a definition are the same thing (that's why they are presented in the same section https://coq.inria.fr/refman/language/core/definitions.html): they associate a name to a term (the body) and its type. If the type is a proposition, then we call the body a "proof term".

When we use a lemma in a proof with `rewrite lemma_name` or `apply lemma_name`, we usually don't care what the proof term (body) of this lemma is. What we care about is its type (the statement of the lemma). In particular, in the case of `apply` the conclusion of the statement (the type thus) should be unified with the current goal, while in the case of `rewrite`, the conclusion of the statement should be an equality (or an equivalence) between two terms, and the left-hand side will be unified with subterms of the current goal.

Both tactics do not actually take `qualid` as argument but terms (`one_term` to be specific on the parsing level). This means that instead of using `lemma_name`, you can also use its proof term directly. But whether you use `lemma_name` or its proof term, what you actually care about is not the term but its type (the statement). This doesn't depend on whether you use a lemma or a term directly.

To take a concrete example, if `foo` is a lemma whose statement (type) is `forall x y z : nat, x = y -> y = z -> x = z`, then `foo a b` is a (proof) term whose type is `forall z : nat, a <= b -> b <= z -> a <= z`, thus you may use to replace a goal of the form `a <= z` with two subgoals `a <= b` and `b <= z`.

Reply:
Most of that makes sense.  A hypothesis has a type but not a value (nor a proof term), right? A bit different from the usual programming language notions.

The last paragraph bears a little more explanation.  The last paragraph is an example of partial application, right?  Is this the only way (or the most common way) to do partial application for a proof term?  Is that the dual form of partial application to a function, e.g.

Fixpoint add2 m := add 2 m.
Compute (add2 5).

We should explain partial application somewhere in the doc but probably not in this file.

Reply:
This didn't work, nor does `(plus_Sn_m 3)`.  Where am I going astray?

```
Print plus_Sn_m.
Goal True.
let t := (plus_Sn_m 3 4) in idtac t.

(* The reference plus_Sn_m was not found in the current environment. *)
```

Reply:
> Most of that makes sense. A hypothesis has a type but not a value (nor a proof term), right?

Right.

>  The last paragraph is an example of partial application, right?

Right.

>  Is this the only way (or the most common way) to do partial application for a proof term? Is that the dual form of partial application to a function

Yes, because these proof terms *are* functions, as `Print` shows.

> This didn't work, nor does `(plus_Sn_m 3)`. Where am I going astray?

Here is your fix: `let t := constr:(plus_Sn_m 3 4) in idtac t.`
The problem was just that the `let` construct of Ltac does not expect a term by default. So it was not looking in the right place of the environment...

Reply:
Thanks.  For `let t := constr:(plus_Sn_m) in idtac t.`, I thought I'd see the proof term.  Am I not seeing it because the proof is opaque?   Is there a way to see it so I can explain it in the text?

Reply:
Yes, for instance `let t := eval unfold plus_Sn_m in plus_Sn_m in idtac t.` fails with `plus_Sn_m is opaque.`

I think only `Print` lets you see through opaque proofs.

Reply:
I'd still like to explain how `rewrite (CR_of_Q_plus R 1 1)` works for this tactic--still don't get it.   It's defined in a Strucure as:

`    CR_of_Q_plus : forall q r : Q, CReq (CR_of_Q (q+r))
                                           (CRplus (CR_of_Q q) (CR_of_Q r));
`

Reply:
Because it's defined in a `Structure` (i.e. a record), it has additional arguments that do not appear on this line (the `R`). Thus `CR_of_Q_plus R 1 1` has the type `CReq (CR_of_Q (1+1)) (CRplus (CR_of_Q 1) (CR_of_Q 1))`. `CReq` is a setoid equality and thus you can rewrite with it.

Reply:
Yeah, that makes sense.  I was thinking that `one_term` had to be an already-proven statement, which is probably wrong--when will `rewrite` introduce introduce `one_term` as a side condition?  And I take it that being in a structure doesn't mean that `CR_of_Q_plus` is proven?

(For "replace", I see the language "This equality is automatically solved if it occurs among the hypotheses, or if its symmetric form occurs.")

Reply:
Even if you did not declare it as a lemma, the simple fact that you can pass around a term means that its type is inhabited and thus proven. Stuff coming from a structure can thus be "proven".

`rewrite` will only generate side-conditions if the lemma that is used has preconditions that are not automatically resolved.

Example:

```coq
Goal 1 = 1.
rewrite <- (min_l 1 1).
```

generates two goals `Nat.min 1 1 = Nat.min 1 1` and the side-condition `1 <= 1`.

For `replace ... with ...`, the side-condition will be the equality itself if it cannot be found in the hypotheses. The confusing bit might be that both `replace` and `rewrite` take a `one_term` as argument but it hasn't the same status at all. In the case of `replace`, it is the subterm to be replaced, while in the case of `rewrite`, it is a proof of an equality or equivalence.

Reply:
> Even if you did not declare it as a lemma, the simple fact that you can pass around a term means that its type is inhabited and thus proven

Ah, OK.  Still need to look at more tactics and consider whether they are using terms or the type of term.  IIRC you said most (all?) tactics are using the type of the term.

How does this work if you pass unprovable terms such as `1 = 0` or `True -> False`?

Reply:
> How does this work if you pass unprovable terms such as 1 = 0 or True -> False?

Their type is `Prop`, which is inhabited.  (Of course it doesn't make sense to rewrite with a lemma which proves that `Prop` is inhabited, so using these particular "proof terms" really only makes sense with something like `pose proof`.)

Reply:
>  IIRC you said most (all?) tactics are using the type of the term.

Most tactics use the term itself to build the proof term, which means that only the type is seen in the goal / context.  The exceptions include `change` (which leaves no evidence in the proof term), `replace` and Ltac `match` (which use the terms you pass them for matching), `pose` (which uses the same term in the proof term evidence and in the body of the variable added to the context), and `assert`/`cut`, which do not take a proof and instead ask you to prove a thing of the type you give them. As usual, I think `refine` is a useful primitive to think about here, since pretty much every other tactic bottoms out in a call to `refine`.